### ISSUE TITLE
Interaction between `LoadCredential` and `ExecStartPost` causing service start failures

### ISSUE THREAD

**Issue by frantisekhanzlikbl** [systemd/systemd#33953]
### systemd version the issue has been seen with

256.2

### Used distribution

NixOS 24.11 (Vicuna)

### Linux kernel version used

6.6.43

### CPU architectures issue was seen on

x86_64

### Component

systemd, systemd-nspawn

### Expected behaviour you didn't see

Service with a `LoadCredential` and `ExecStartPost` settings should start and work normally.

### Unexpected behaviour you saw

Service with a `LoadCredential` and `ExecStartPost` settings non-deterministically, but often (about one fourth of the time) refuses to start and work normally. The journal contains the following messages:
```log
my_test.service: Failed to set up credentials: Protocol error
my_test.service: Failed at step CREDENTIALS spawning /usr/bin/env: Protocol error
```

### Steps to reproduce the problem

1. Create an arbitrary file at `/tmp/my_secret`
2. Create a system service with the following definition:
    ```ini
    [Unit]
  
    [Service]
    ExecStart=/usr/bin/env true
    ExecStartPost=/usr/bin/env true
    LoadCredential=secret:/tmp/my_secret
    ```
3. Restart the service a few times via `systemctl restart my_test.service`.

I'm not sure why exactly does using `ExecStartPost` matter, but the issue is not reproducible for me without it. The actual commands used as part of `ExecStart` or `ExecStartPost` do not seem to be important.

I am aware of https://github.com/systemd/systemd/pull/31197, but since it targets a slightly different issue and I don't have experience with upgrading systemd beyond what is packaged on NixOS master, I decided to report this issue as is. Please let me know if I should attempt to test on 256.4.

### Additional program output to the terminal or log subsystem illustrating the issue

```sh
Aug 06 20:04:10 tartaros systemd[1]: my_test.service: Trying to enqueue job my_test.service/restart/replace
Aug 06 20:04:10 tartaros systemd[1]: my_test.service: Installed new job my_test.service/restart as 165071
Aug 06 20:04:10 tartaros systemd[1]: my_test.service: Enqueued job my_test.service/restart as 165071
Aug 06 20:04:10 tartaros systemd[1]: my_test.service: Job 165071 my_test.service/restart finished, result=done
Aug 06 20:04:10 tartaros systemd[1]: my_test.service: Converting job my_test.service/restart -> my_test.service/start
Aug 06 20:04:10 tartaros systemd[1]: my_test.service: Will spawn child (service_enter_start): /usr/bin/env
Aug 06 20:04:10 tartaros systemd[1]: my_test.service: Failed to set 'memory.zswap.writeback' attribute on '/system.slice/my_test.service' to '1': No such file or directory
Aug 06 20:04:10 tartaros systemd[1]: my_test.service: Passing 0 fds to service
Aug 06 20:04:10 tartaros systemd[1]: my_test.service: About to execute: /usr/bin/env true
Aug 06 20:04:10 tartaros systemd[1]: my_test.service: Forked /usr/bin/env as 1566820 (via CLONE_INTO_CGROUP)
Aug 06 20:04:10 tartaros systemd[1]: my_test.service: Will spawn child (service_enter_start_post): /usr/bin/env
Aug 06 20:04:10 tartaros systemd[1]: my_test.service: About to execute: /usr/bin/env true
Aug 06 20:04:10 tartaros systemd[1]: my_test.service: Forked /usr/bin/env as 1566821 (via CLONE_INTO_CGROUP)
Aug 06 20:04:10 tartaros systemd[1]: my_test.service: Changed dead -> start-post
Aug 06 20:04:10 tartaros systemd[1]: Starting my_test.service...
Aug 06 20:04:10 tartaros (env)[1566820]: Found cgroup2 on /sys/fs/cgroup/, full unified hierarchy
Aug 06 20:04:10 tartaros (env)[1566821]: Found cgroup2 on /sys/fs/cgroup/, full unified hierarchy
Aug 06 20:04:10 tartaros (env)[1566820]: Found cgroup2 on /sys/fs/cgroup/, full unified hierarchy
Aug 06 20:04:10 tartaros (env)[1566821]: Found cgroup2 on /sys/fs/cgroup/, full unified hierarchy
Aug 06 20:04:10 tartaros (env)[1566820]: Successfully forked off '(sd-mkdcreds)' as PID 1566822.
Aug 06 20:04:10 tartaros (env)[1566821]: Successfully forked off '(sd-mkdcreds)' as PID 1566823.
Aug 06 20:04:10 tartaros (sd-m[1566822]: Changing mount propagation /dev (MS_REC|MS_SLAVE "")
Aug 06 20:04:10 tartaros (sd-m[1566822]: Umounting /run/credentials/my_test.service...
Aug 06 20:04:10 tartaros (sd-m[1566822]: Mounting tmpfs (tmpfs) on /dev/shm (MS_NOSUID|MS_NODEV|MS_NOEXEC|MS_NOSYMFOLLOW "mode=0700,nr_inodes=1024,size=1048576,noswap")...
Aug 06 20:04:10 tartaros (env)[1566821]: (sd-mkdcreds) succeeded.
Aug 06 20:04:10 tartaros (sd-m[1566822]: Changing mount flags /dev/shm (MS_RDONLY|MS_NOSUID|MS_NODEV|MS_NOEXEC|MS_REMOUNT|MS_NOSYMFOLLOW|MS_BIND "")...
Aug 06 20:04:10 tartaros (sd-m[1566822]: Moving mount /dev/shm → /run/credentials/my_test.service (MS_MOVE "")...
Aug 06 20:04:10 tartaros (sd-m[1566822]: Failed to mount /dev/shm (type n/a) on /run/credentials/my_test.service (MS_MOVE ""): No such file or directory
Aug 06 20:04:10 tartaros (env)[1566821]: my_test.service: Executing: /usr/bin/env true
Aug 06 20:04:10 tartaros (env)[1566820]: (sd-mkdcreds) failed with exit status 1.
Aug 06 20:04:10 tartaros (env)[1566820]: my_test.service: Failed to set up credentials: Protocol error
Aug 06 20:04:10 tartaros (env)[1566820]: my_test.service: Failed at step CREDENTIALS spawning /usr/bin/env: Protocol error
Aug 06 20:04:10 tartaros systemd[1]: my_test.service: Child 1566820 belongs to my_test.service.
Aug 06 20:04:10 tartaros systemd[1]: my_test.service: Main process exited, code=exited, status=243/CREDENTIALS
Aug 06 20:04:10 tartaros systemd[1]: my_test.service: Child 1566821 belongs to my_test.service.
Aug 06 20:04:10 tartaros systemd[1]: my_test.service: Control process exited, code=exited, status=0/SUCCESS (success)
Aug 06 20:04:10 tartaros systemd[1]: my_test.service: Got final SIGCHLD for state start-post.
Aug 06 20:04:10 tartaros systemd[1]: my_test.service: Failed with result 'exit-code'.
Aug 06 20:04:10 tartaros systemd[1]: my_test.service: Service will not restart (restart setting)
Aug 06 20:04:10 tartaros systemd[1]: my_test.service: Changed start-post -> failed
Aug 06 20:04:10 tartaros systemd[1]: my_test.service: Job 165071 my_test.service/start finished, result=failed
Aug 06 20:04:10 tartaros systemd[1]: Failed to start my_test.service.
Aug 06 20:04:10 tartaros systemd[1]: my_test.service: Unit entered failed state.
Aug 06 20:04:10 tartaros systemd[1]: my_test.service: Consumed 10ms CPU time, 3.3M memory peak.
Aug 06 20:04:10 tartaros systemd[1]: my_test.service: Releasing resources...
```

**Comment by YHNdnzj** [comment]
Should be fixed by #32612.

**Comment by frantisekhanzlikbl** [comment]
Yep, using `Type = exec` has fixed the issue.

An interesting thing is that after testing this more, I could reproduce even without `ExecStartPost`, albeit much less often.

Sorry for the wrong label (I mixed up systemd-nspawn and systemd-run in my head) and thanks so much for the immediate reply! :slightly_smiling_face:
